### PR TITLE
Do proper reference counting with callouts

### PR DESCRIPTION
**Pull request by taylor-b**
This handles the race condition between a timer executing and being deleted.

sctp_timer_start increments reference counts, and the timer executing or sctp_timer_stop being called decrements them.

### CODE REVIEW

**Comment by taylor-b** [comment]
Note that it would be

File: usrsctplib/netinet/sctp_callout.c [hunk 1]
```diff
@@ -77,60 +79,28 @@ uint32_t sctp_get_tick_count(void) {
...
  * - SCTP_BASE_INFO(callqueue)
  * - sctp_os_timer_next: next timer to check
- * - sctp_os_timer_current: current callout callback in progress
- * - sctp_os_timer_current_tid: current callout thread id in progress
- * - sctp_os_timer_waiting: some thread is waiting for callout to complete
- * - sctp_os_timer_wait_ctr: incremented every time a thread wants to wait
- *                           for a callout to complete.
  */
 static sctp_os_timer_t *sctp_os_timer_next = NULL;
-static sctp_os_timer_t *sctp_os_timer_current = NULL;
```
Comment (taylor-b) [inline usrsctplib/netinet/sctp_callout.c]: Note that this and a lot of stuff below is just reverting the problematic commits that cause the deadlock. Again, look at individual commits for clearer diff.

File: usrsctplib/netinet/sctp_callout.c [hunk 2]
```diff
@@ -77,60 +79,28 @@ uint32_t sctp_get_tick_count(void) {
...
-
 
 void
 sctp_os_timer_init(sctp_os_timer_t *c)
 {
 	memset(c, 0, sizeof(*c));
 }
 
-void
+int
```
Comment (taylor-b) [inline usrsctplib/netinet/sctp_callout.c]: This return value is consistent with callout_reset, which is what SCTP_OS_TIMER_START maps to in FreeBSD.

File: usrsctplib/netinet/sctp_pcb.c [hunk 3]
```diff
@@ -4151,10 +4149,10 @@ sctp_inpcb_free(struct sctp_inpcb *inp, int immediate, int from)
 	if (SCTP_ASOC_CREATE_LOCK_CONTENDED(inp))
 		being_refed++;
 
+	// NOTE: 0 refcount also means no timers are referencing us.
 	if ((inp->refcount) ||
 	    (being_refed) ||
 	    (inp->sctp_flags & SCTP_PCB_FLAGS_CLOSE_IP)) {
-		(void)SCTP_OS_TIMER_STOP(&inp->sctp_ep.signature_change.timer);
```
Comment (taylor-b) [inline usrsctplib/netinet/sctp_pcb.c]: This was already stopped above, so I don't think it needs to be called again?

File: usrsctplib/netinet/sctp_pcb.c [hunk 4]
```diff
@@ -4173,19 +4171,6 @@ sctp_inpcb_free(struct sctp_inpcb *inp, int immediate, int from)
 	SCTP_INP_WUNLOCK(inp);
 	SCTP_ASOC_CREATE_UNLOCK(inp);
 	SCTP_INP_INFO_WUNLOCK();
-	/* Now we release all locks. Since this INP
```
Comment (taylor-b) [inline usrsctplib/netinet/sctp_pcb.c]: No longer necessary since at this point we are assured no timer referencing us is running.

File: usrsctplib/netinet/sctp_indata.c [hunk 5]
```diff
@@ -1745,7 +1745,7 @@ sctp_process_a_data_chunk(struct sctp_tcb *stcb, struct sctp_association *asoc,
 		 * empty data chunk.
 		 */
 		op_err = sctp_generate_no_user_data_cause(tsn);
-		stcb->sctp_ep->last_abort_code = SCTP_FROM_SCTP_INDATA + SCTP_LOC_14;
+		stcb->sctp_ep->last_abort_code = SCTP_FROM_SCTP_INDATA + SCTP_LOC_15;
```
Comment (mstyura) [inline usrsctplib/netinet/sctp_indata.c]: @taylor-b is this change done to resolve uniqueness violation, because `SCTP_LOC_14` is used here:
https://github.com/sctplab/usrsctp/blob/32020d6f44d405ef4ea3fecb52bc8a392aa8a6c4/usrsctplib/netinet/sctp_indata.c#L1592
?

File: usrsctplib/netinet/sctp_pcb.c (hunk 3 above)
Comment (mstyura) [reply usrsctplib/netinet/sctp_pcb.c]: Note: the place where it stopped is here:

https://github.com/sctplab/usrsctp/blob/0f1de1e6896523ac790ca0880fd6c91dc8a3d7cb/usrsctplib/netinet/sctp_pcb.c#L3932-L3933

File: usrsctplib/netinet/sctp_pcb.c [hunk 6]
```diff
@@ -5734,11 +5737,9 @@ sctp_free_assoc(struct sctp_inpcb *inp, struct sctp_tcb *stcb, int from_inpcbfre
 	 */
 	(void)SCTP_OS_TIMER_STOP(&asoc->strreset_timer.timer);
 	(void)SCTP_OS_TIMER_STOP(&asoc->dack_timer.timer);
-	(void)SCTP_OS_TIMER_STOP(&asoc->strreset_timer.timer);
```
Comment (mstyura) [inline usrsctplib/netinet/sctp_pcb.c]: As far I understand here `SCTP_OS_TIMER_STOP` is used and not `sctp_timer_stop` because otherwise reference counter would be decremented twice, where first time it was decremented here:
https://github.com/sctplab/usrsctp/blob/0f1de1e6896523ac790ca0880fd6c91dc8a3d7cb/usrsctplib/netinet/sctp_pcb.c#L5555-L5568 is that correct?

File: usrsctplib/netinet/sctp_pcb.c [hunk 7]
```diff
@@ -5533,36 +5534,38 @@ sctp_free_assoc(struct sctp_inpcb *inp, struct sctp_tcb *stcb, int from_inpcbfre
...
-		net->hb_timer.self = NULL;
+		sctp_timer_stop(SCTP_TIMER_TYPE_SEND, inp, stcb, net,
+		                SCTP_FROM_SCTP_PCB + SCTP_LOC_22);
+		sctp_timer_stop(SCTP_TIMER_TYPE_INIT, inp, stcb, net,
+		                SCTP_FROM_SCTP_PCB + SCTP_LOC_23);
+		sctp_timer_stop(SCTP_TIMER_TYPE_SHUTDOWN, inp, stcb, net,
+		                SCTP_FROM_SCTP_PCB + SCTP_LOC_24);
+		sctp_timer_stop(SCTP_TIMER_TYPE_COOKIE, inp, stcb, net,
+		                SCTP_FROM_SCTP_PCB + SCTP_LOC_25);
+		sctp_timer_stop(SCTP_TIMER_TYPE_SHUTDOWNACK, inp, stcb, net,
```
Comment (mstyura) [inline usrsctplib/netinet/sctp_pcb.c]: As I understand `SCTP_TIMER_TYPE_SEND`, `SCTP_TIMER_TYPE_INIT`, `SCTP_TIMER_TYPE_SHUTDOWN`, `SCTP_TIMER_TYPE_COOKIE`, `SCTP_TIMER_TYPE_SHUTDOWNACK` all maps to `net->rxt_timer`, so it means that ref counter would be decremented 5 times. Is that correct that existing code already properly used `sctp_timer_start` which does increment?

File: usrsctplib/netinet/sctputil.c [hunk 8]
```diff
@@ -2529,7 +2537,21 @@ sctp_timer_stop(int t_type, struct sctp_inpcb *inp, struct sctp_tcb *stcb,
...
+		 * incremented in sctp_timer_start.
+		 */
+		if (inp) {
+			SCTP_INP_DECR_REF(inp);
+		}
+		if (stcb) {
+			atomic_add_int(&stcb->asoc.refcnt, -1);
+		}
+		if (net) {
+			sctp_free_remote_addr(net);
```
Comment (mstyura) [inline usrsctplib/netinet/sctputil.c]: Not really sure about this, but better to clarify rather than ignore: shouldn't reference be decremented in the reverse order of being incremented?

File: usrsctplib/netinet/sctp_pcb.c (hunk 6 above)
Comment (mstyura) [reply usrsctplib/netinet/sctp_pcb.c]: BTW, other than this block of `SCTP_OS_TIMER_STOP` usage there one more usage here:

https://github.com/sctplab/usrsctp/blob/0f1de1e6896523ac790ca0880fd6c91dc8a3d7cb/usrsctplib/netinet/sctp_pcb.c#L8072

Shouldn't `sctp_timer_stop` be used here or at least comment why not?

File: usrsctplib/netinet/sctp_indata.c (hunk 5 above)
Comment (taylor-b) [reply usrsctplib/netinet/sctp_indata.c]: Yes, as long as I was adding more SCTP_LOCs I went ahead and made the rest ordered and unique.

File: usrsctplib/netinet/sctp_pcb.c (hunk 6 above)
Comment (taylor-b) [reply usrsctplib/netinet/sctp_pcb.c]: That's correct. And yikes, can't believe I missed that, good catch. Fixed.

File: usrsctplib/netinet/sctp_pcb.c (hunk 7 above)
Comment (taylor-b) [reply usrsctplib/netinet/sctp_pcb.c]: It won't decrement five times, because the timer can only be in the list once, with one type, so only the call with a matching type will cause a decrement. It's unfortunate that one function call turned into five, but I didn't want to make a separate timer stop function that takes the timer as an argument just for this. Though maybe I should?

And yes, sctp_timer_start was already being called everywhere.

File: usrsctplib/netinet/sctputil.c (hunk 8 above)
Comment (taylor-b) [reply usrsctplib/netinet/sctputil.c]: It doesn't really make a difference here, but I could change it for stylistic purposes if people want.

File: usrsctplib/netinet/sctp_pcb.c (hunk 7 above)
Comment (mstyura) [reply usrsctplib/netinet/sctp_pcb.c]: > Though maybe I should?

IMO, no. Just wanted to clarify the behavior here.

File: usrsctplib/netinet/sctputil.c [hunk 9]
```diff
@@ -1686,58 +1732,33 @@ sctp_timeout_handler(void *t)
 		     (tmr->type != SCTP_TIMER_TYPE_SHUTDOWNACK) &&
 		     (tmr->type != SCTP_TIMER_TYPE_SHUTDOWNGUARD) &&
 		     (tmr->type != SCTP_TIMER_TYPE_ASOCKILL))) {
-			SCTP_INP_DECR_REF(inp);
```
Comment (mstyura) [inline usrsctplib/netinet/sctputil.c]: This decrement migrated to here as I understand:
https://github.com/sctplab/usrsctp/blob/4b77ee20ffa7a2b684f48e802f7b529a7151d978/usrsctplib/netinet/sctputil.c#L2584-L2598

**Comment by mstyura** [comment]
Reviewed latest version of this PR. Looks good to me.

**Comment by mstyura** [comment]
I've just realized that there is one more direct usage of `sctp_timer_stop` caused by usage of [`SCTP_OS_TIMER_STOP_DRAIN`](https://github.com/taylor-b/usrsctp/blob/4b77ee20ffa7a2b684f48e802f7b529a7151d978/usrsctplib/netinet/sctp_callout.h#L106). 
https://github.com/sctplab/usrsctp/blob/4b77ee20ffa7a2b684f48e802f7b529a7151d978/usrsctplib/netinet/sctp_pcb.c#L6943
which could possibly replaced with:
```
sctp_timer_stop(SCTP_TIMER_TYPE_ADDR_WQ,
            ???,
            ???,
            ???);
```
It will probably does not affect reference counting, but it might make sense to change it as well (and remove `SCTP_OS_TIMER_STOP_DRAIN` completely).

**Comment by taylor-b** [comment]
@mstyura This usage is okay because it doesn't involve any reference counted structures. And given everything that's about to be freed I assumed the "stop drain" behavior is desired here.

**Comment by mstyura** [comment]
@tuexen hello prof. Michael Tüxen!
Could you please review this pull request?
Thanks a lot in advance!

**Comment by tuexen** [comment]
> @tuexen hello prof. Michael Tüxen!
> Could you please review this pull request?
> Thanks a lot in advance!

It is on my agenda for next week...

**Comment by lgrahl** [comment]
It's great to see you back, @taylor-b!

**Comment by tuexen** [comment]
Hi @taylor-b ,

is it OK for you if I cherry pick the cleanup changes from the PR, commit them, and then you rebase the patch? Then we can easily focus on non-cleanup changes.

**Comment by taylor-b** [comment]
@tuexen Yeah, that's fine.

**Comment by taylor-b** [comment]
@tuexen Rebased.

**Comment by tuexen** [comment]
@taylor-b Can you rebase (again) and avoid comments of the form `// ...`. Please replace them ny using `/* ... */`. The remaining changes should be the actual fix.

**Comment by taylor-b** [comment]
@tuexen Done.

**Comment by tuexen** [comment]
> @tuexen Done.

Thanks. Will review and test it also in the FreeBSD kernel stack.

File: usrsctplib/netinet/sctp_pcb.c [hunk 10]
```diff
@@ -5237,6 +5225,21 @@ sctp_remove_net(struct sctp_tcb *stcb, struct sctp_nets *net)
...
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_11);
+	sctp_timer_stop(SCTP_TIMER_TYPE_COOKIE, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_12);
+	sctp_timer_stop(SCTP_TIMER_TYPE_SHUTDOWNACK, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_13);
+	sctp_timer_stop(SCTP_TIMER_TYPE_PATHMTURAISE, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_14);
+	sctp_timer_stop(SCTP_TIMER_TYPE_HEARTBEAT, inp, stcb, net,
+	                SCTP_FROM_SCTP_PCB + SCTP_LOC_15);
+	net->removed = 1; /* Prevents timers from being restarted. */
```
Comment (tuexen) [inline usrsctplib/netinet/sctp_pcb.c]: I think stopping the HB and PMTU timer here is actually the correct place. However, we should not stop the other timers here, since, if for example the SEND timer is not running anymore, no retransmissions for the DATA chunks sent on the now removed net will happen. I committed a corresponding fix.

File: usrsctplib/netinet/sctp_pcb.c [hunk 11]
```diff
@@ -5531,35 +5534,8 @@ sctp_free_assoc(struct sctp_inpcb *inp, struct sctp_tcb *stcb, int from_inpcbfre
...
-	asoc->dack_timer.self = NULL;
-	(void)SCTP_OS_TIMER_STOP(&asoc->strreset_timer.timer);
-	/*-
-	 * For stream reset we don't blast this unless
-	 * it is a str-reset timer, it might be the
-	 * free-asoc timer which we DON'T want to
-	 * disturb.
-	 */
-	if (asoc->strreset_timer.type == SCTP_TIMER_TYPE_STRRESET)
-		asoc->strreset_timer.self = NULL;
```
Comment (tuexen) [inline usrsctplib/netinet/sctp_pcb.c]: Since the assoc kill timer is kept in place, I added a boolean variable to `sctp_stop_association_timers()`.

**Comment by tuexen** [comment]
I committed all the timer code cleanup to reduce the diff for the fix. I also took stopping the HB and PMTU timer an net removal. I think we can't stop the others...

Can you rebase? The diff should be much smaller now.

File: usrsctplib/netinet/sctp_pcb.c (hunk 10 above)
Comment (taylor-b) [reply usrsctplib/netinet/sctp_pcb.c]: Can you summarize how retransmissions work? I'm now worried that the `net->removed` logic is also flawed; should it be possible to restart the SEND timer after removing?

I'm doing this because the old behavior was "stop timers when there are zero references". This doesn't work since the timers themselves now add references.

So I assumed (pretty naively I now realize) that I could just stop the timers here. If that's not possible, then we still have the same issue: a timer holding a reference could keep the net alive when it otherwise would have been freed...

If we want exactly the same behavior as before, maybe we could use two reference counts as I mentioned previously.

When the main reference count is reduced to 0, we'd stop the timers (as they were stopped before), but would only free the memory if the timer reference count is also 0.

The new problem would be that we'd be setting the `removed`/"about to be freed" flag inside `sctp_free_remote_addr`, which would be racy since I'm assuming there is no guarantee that the association lock is held wherever that's called.

So I'm somewhat at a loss now. We could address that problem by adding a `NETKILL` timer but now things are getting pretty messy.

I went ahead and rebased, but I want to be sure this is watertight before merging. Do you have any ideas?

File: usrsctplib/netinet/sctp_callout.c [hunk 12]
```diff
@@ -49,10 +49,12 @@
 #include <user_atomic.h>
 #include <netinet/sctp_sysctl.h>
 #include <netinet/sctp_pcb.h>
+#include <netinet/sctp_var.h>
```
Comment (tuexen) [inline usrsctplib/netinet/sctp_callout.c]: I think this is not needed. Can you remove it?

File: usrsctplib/netinet/sctp_callout.c [hunk 13]
```diff
@@ -49,10 +49,12 @@
...
 #else
 #include <netinet/sctp_os.h>
 #include <netinet/sctp_callout.h>
 #include <netinet/sctp_pcb.h>
+#include <netinet/sctp_var.h>
```
Comment (tuexen) [inline usrsctplib/netinet/sctp_callout.c]: I think this is not needed. Can you remove it?

File: usrsctplib/netinet/sctp_pcb.c (hunk 10 above)
Comment (tuexen) [reply usrsctplib/netinet/sctp_pcb.c]: > Can you summarize how retransmissions work? I'm now worried that the `net->removed` logic is also flawed; should it be possible to restart the SEND timer after removing?

If you send a DATA chunk on a `net`, you store a pointer to it in `whoTo` and start a timer for it. Assuming we have multiple `net`s, when the timer runs off, we choose a different net, retransmit the DATA chunk there and start a new timer on the other `net`.

If we remove a `net`, and there is outstanding data, we need to wait until the timer runs off. In the old code, we would not free the net since there was a ref count on all outstanding DATA chunks. But I think this should work with the new stuff the same way.

I guess this can be optimized be moving chunks over to a different `net`, but that can be a separate commit.
This also does not apply to the WebRTC case, since dynamic address configuration is not used there.
> 
> I'm doing this because the old behavior was "stop timers when there are zero references". This doesn't work since the timers themselves now add references.
We would remove the net, after the timer is running off. That is the same as in the old case, since we would have ref counts on the DATA chunks.
> 
> So I assumed (pretty naively I now realize) that I could just stop the timers here. If that's not possible, then we still have the same issue: a timer holding a reference could keep the net alive when it otherwise would have been freed...

I don't see that as an issue and the behaviour should be the same.

> 
> If we want exactly the same behavior as before, maybe we could use two reference counts as I mentioned previously.
> 
> When the main reference count is reduced to 0, we'd stop the timers (as they were stopped before), but would only free the memory if the timer reference count is also 0.

When we are removing the `net`, it does not make any sense to send packets via the `net`. So we should stop the HB and PMTU timer right away. That is why I committed that. The DATA retransmission must be kept running... Or me move the data away...

> 
> The new problem would be that we'd be setting the `removed`/"about to be freed" flag inside `sctp_free_remote_addr`, which would be racy since I'm assuming there is no guarantee that the association lock is held wherever that's called.

I think the STCB lock should be hold when modifying any `net`. Do you have a place in mind, where it is not held?

> 
> So I'm somewhat at a loss now. We could address that problem by adding a `NETKILL` timer which would repeatedly attempt to stop the timers, but now things are getting pretty messy.
> 
> I went ahead and rebased, but I want to be sure this is watertight before merging. Do you have any ideas?

Thanks for re-basing. I'm trying to commit things in small pieces to reduce the diff and get a better view regarding the the changes. I think ref counting should work. I'm also putting the changes to the FreeBSD kernel, since we have there some features for detecting memory loss or locking issues...

**Comment by tuexen** [comment]
I replaced the `bool` by an `int` as you suggested but used the `SCTP_ADDR_BEING_DELETED` flag of the `dest_state` field` instead of adding a `removed` field to the `net` structure. After the removal of `#include <netinet/sctp_var.h>`, could you rebase again?

File: usrsctplib/netinet/sctp_pcb.c (hunk 10 above)
Comment (taylor-b) [reply usrsctplib/netinet/sctp_pcb.c]: >  Do you have a place in mind, where it is not held?
No, I just wasn't sure and didn't look exhaustively. I assumed that since the atomic increment was being used, that was the only thing guaranteeing thread safety.

Thanks for your explanation and for bearing with me.

File: usrsctplib/netinet/sctp_callout.c (hunk 13 above)
Comment (taylor-b) [reply usrsctplib/netinet/sctp_callout.c]: Done, and rebased.

**Comment by taylor-b** [comment]
@tuexen Ping

**Comment by tuexen** [comment]
@taylor-b Still working on some cleanup of the timer code...

**Comment by tuexen** [comment]
@taylor-b Just to understand your concept crystal clear:

- Whenever you start a timer, increment refcounts to whatever is referenced (inp, stcb, net).
- Whenever you stop a running timer, decrement refcounts to whatever is referenced (inp, stcb, net).
- Whenever a timer runs off, decrement refcounts to whatever is referenced (inp, stcb, net).

Here is an overview of the various timers:

|Timer				|Object				|ref inp	|ref stcb	|ref net	|
|-------------------------------|-------------------------------|-------|-------|-------|
|SCTP_TIMER_TYPE_SEND		|net->rxt_timer			|Yes	|Yes	|Yes	|
|SCTP_TIMER_TYPE_INIT		|net->rxt_timer			|Yes	|Yes	|Yes	|
|SCTP_TIMER_TYPE_RECV		|stcb->asoc.dack_timer		|Yes	|Yes	|No	|
|SCTP_TIMER_TYPE_SHUTDOWN	|net->rxt_timer			|Yes	|Yes	|Yes	|
|SCTP_TIMER_TYPE_HEARTBEAT	|net->hb_timer			|Yes	|Yes	|Yes	|
|SCTP_TIMER_TYPE_COOKIE		|net->rxt_timer			|Yes	|Yes	|Yes	|
|SCTP_TIMER_TYPE_NEWCOOKIE	|inp->sctp_ep.signature_change	|Yes	|No	|No	|
|SCTP_TIMER_TYPE_PATHMTURAISE	|net->pmtu_timer		|Yes	|Yes	|Yes	|
|SCTP_TIMER_TYPE_SHUTDOWNACK	|net->rxt_timer			|Yes	|Yes	|Yes	|
|SCTP_TIMER_TYPE_ASCONF		|stcb->asoc.asconf_timer	|Yes	|Yes	|Yes	|
|SCTP_TIMER_TYPE_SHUTDOWNGUARD	|stcb->asoc.shut_guard_timer	|Yes	|Yes	|No	|
|SCTP_TIMER_TYPE_AUTOCLOSE	|stcb->asoc.autoclose_timer	|Yes	|Yes	|No	|
|SCTP_TIMER_TYPE_STRRESET	|stcb->asoc.strreset_timer	|Yes	|Yes	|Yes	|
|SCTP_TIMER_TYPE_INPKILL	|inp->sctp_ep.signature_change	|Yes	|No	|No	|
|SCTP_TIMER_TYPE_ASOCKILL	|stcb->asoc.strreset_timer	|Yes	|Yes	|No	|
|SCTP_TIMER_TYPE_ADDR_WQ	|SCTP_BASE_INFO(addr_wq_timer)	|No	|No	|No	|
|SCTP_TIMER_TYPE_PRIM_DELETED	|stcb->asoc.delete_prim_timer	|Yes	|Yes	|No	|

**Comment by taylor-b** [comment]
@tuexen That's correct. It follows the pattern of "if you reference something but don't hold its lock, you need to hold a reference," which is what I saw the rest of the codebase doing.

**Comment by mstyura** [comment]
Hello @tuexen and @taylor-b,
Hope you and your families feel well and safe!
I've got slightly lost on the status of this PR. I've seen many parts from PR was kind of "cherry picked" into master already, but some of the changes are still not merged. Are they pending further work or abandoned due to not being necessary anymore? I see that Chromium switched to some commit which is no longer directly reachable from `master` probably due to force push happen... @taylor-b maybe it make sense to adjust commit in `DEPS` in `chromium` and `webrtc` as well?

Thanks a lot in advance!

**Comment by tuexen** [comment]
@mstyura I'm in the process of getting it in. The current state is that I cleaned up the timer code to avoid cases where we would increment/decrement in a way we shouldn't. We now have KASSERTs in place and some fuzz tester are finding issues (syzkaller based). Once these bugs are  fixed, adding the refcounts is a simple change. That is what I'm working towards.

**Comment by mstyura** [comment]
Thank you very much @tuexen!

**Comment by Lastique** [comment]
If this PR is still relevant, could it be rebased on the current master please?

**Comment by tuexen** [comment]
> If this PR is still relevant, could it be rebased on the current master please?

It is still relevant. I'm still cleaning up the counters before making use of them in this context.

**Comment by tuexen** [comment]
A modified version of this PR has been committed in https://github.com/sctplab/usrsctp/commit/939d48f9632d69bf170c7a84514b312b6b42257d.

**Comment by taylor-b** [comment]
In that case, closing.
